File: internal/consolidation/service.go
// Package consolidation provides memory consolidation services
package consolidation

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/cortex-ai/cortex-ai/internal/config"
	"github.com/cortex-ai/cortex-ai/internal/memory"
	"github.com/cortex-ai/cortex-ai/internal/storage"
	"github.com/google/uuid"
)

// Service handles memory consolidation operations
type Service struct {
	storage  storage.Storage
	embedder memory.Embedder
	config   *config.ConsolidationConfig
}

// NewService creates a new consolidation service
func NewService(store storage.Storage, embedder memory.Embedder, cfg *config.ConsolidationConfig) *Service {
	return &Service{
		storage:  store,
		embedder: embedder,
		config:   cfg,
	}
}

// Consolidate processes input and creates or merges a memory
func (s *Service) Consolidate(ctx context.Context, input memory.ConsolidateInput) (*memory.ConsolidateResult, error) {
	// Validate input
	if err := input.Validate(); err != nil {
		return nil, fmt.Errorf("invalid input: %w", err)
	}

	// Set default source if not provided
	if input.Context.Source == "" {
		input.Context.Source = "manual"
	}

	// Set timestamp if not provided
	if input.Context.Timestamp.IsZero() {
		input.Context.Timestamp = time.Now()
	}

	// Generate embedding for the synthesis
	embedding, err := s.embedder.Embed(ctx, input.Synthesis)
	if err != nil {
		return nil, fmt.Errorf("failed to generate embedding: %w", err)
	}

	// If force is set, skip duplicate check
	if !input.Force {
		searchOpts := memory.SearchOptions{
			TopK:         1,
			MinScore:     s.config.SimilarityThreshold,
			FilterLevels: []memory.MemoryLevel{input.Level},
		}
		if input.Level == memory.MemoryLevelWorking {
			searchOpts.SessionID = input.Context.SessionID
		}
		results, err := s.storage.SearchAllLayers(ctx, embedding, searchOpts)
		if err != nil {
			return nil, fmt.Errorf("failed to find similar memories: %w", err)
		}
		if len(results) > 0 {
			return s.mergeWithExisting(ctx, results[0].Memory, input, embedding)
		}
	}

	// Create new memory
	return s.createNew(ctx, input, embedding)
}

func (s *Service) createNew(ctx context.Context, input memory.ConsolidateInput, embedding []float64) (*memory.ConsolidateResult, error) {
	now := time.Now()
	mem := &memory.Memory{
		ID:        uuid.New().String(),
		Level:     input.Level,
		Title:     deriveTitle(input.Synthesis),
		Content:   input.Synthesis,
		Tags:      input.Context.Tags,
		Embedding: embedding,
		Context:   input.Context,
		CreatedAt: now,
		UpdatedAt: now,
		Obsolete:  false,
	}

	if err := mem.Validate(); err != nil {
		return nil, fmt.Errorf("invalid memory: %w", err)
	}

	if err := s.storage.Save(ctx, mem); err != nil {
		return nil, fmt.Errorf("failed to save memory: %w", err)
	}

	return &memory.ConsolidateResult{
		Action:   "created",
		MemoryID: mem.ID,
		Reason:   "new memory created",
	}, nil
}

func (s *Service) mergeWithExisting(ctx context.Context, existing *memory.Memory, input memory.ConsolidateInput, newEmbedding []float64) (*memory.ConsolidateResult, error) {
	// Merge content - append new synthesis to existing
	mergedContent := existing.Content + "\n\n---\n\n" + input.Synthesis

	// Generate new embedding for merged content
	mergedEmbedding, err := s.embedder.Embed(ctx, mergedContent)
	if err != nil {
		// Fall back to using the new embedding
		mergedEmbedding = newEmbedding
	}

	// Update existing memory
	existing.Content = mergedContent
	existing.Embedding = mergedEmbedding
	existing.UpdatedAt = time.Now()

	if len(input.Context.Tags) > 0 {
		existing.Tags = mergeTags(existing.Tags, input.Context.Tags)
		existing.Context.Tags = mergeTags(existing.Context.Tags, input.Context.Tags)
	}

	if len(input.Context.RelatedMemories) > 0 {
		existing.Context.RelatedMemories = mergeTags(existing.Context.RelatedMemories, input.Context.RelatedMemories)
	}

	if err := s.storage.Update(ctx, existing); err != nil {
		return nil, fmt.Errorf("failed to update memory: %w", err)
	}

	return &memory.ConsolidateResult{
		Action:     "merged",
		MemoryID:   existing.ID,
		MergedWith: existing.ID,
		Reason:     fmt.Sprintf("merged with existing memory (similarity >= %.2f)", s.config.SimilarityThreshold),
	}, nil
}

// TransferWorkingToEpisodic transfers working memories to episodic level
func (s *Service) TransferWorkingToEpisodic(ctx context.Context, sessionID string) (int, error) {
	return s.storage.TransferWorkingToEpisodic(ctx, sessionID)
}

// PromoteToSemantic promotes an episodic memory to semantic level
func (s *Service) PromoteToSemantic(ctx context.Context, memoryID string, newContent string) (*memory.ConsolidateResult, error) {
	existing, err := s.storage.Get(ctx, memoryID)
	if err != nil {
		return nil, fmt.Errorf("failed to get memory: %w", err)
	}

	if existing.Level != memory.MemoryLevelEpisodic {
		return nil, fmt.Errorf("can only promote episodic memories to semantic")
	}

	content := newContent
	if content == "" {
		content = existing.Content
	}

	embedding, err := s.embedder.Embed(ctx, content)
	if err != nil {
		return nil, fmt.Errorf("failed to generate embedding: %w", err)
	}

	semantic := &memory.Memory{
		ID:         uuid.New().String(),
		Level:      memory.MemoryLevelSemantic,
		Title:      existing.Title,
		Content:    content,
		Tags:       existing.Tags,
		Embedding:  embedding,
		Context:    existing.Context,
		CreatedAt:  time.Now(),
		UpdatedAt:  time.Now(),
		MergedFrom: []string{memoryID},
		Obsolete:   false,
	}
	semantic.Context.Source = "manual"

	if err := s.storage.Save(ctx, semantic); err != nil {
		return nil, fmt.Errorf("failed to save semantic memory: %w", err)
	}

	return &memory.ConsolidateResult{
		Action:     "created",
		MemoryID:   semantic.ID,
		MergedWith: memoryID,
		Reason:     "promoted from episodic to semantic",
	}, nil
}

// Get retrieves a memory by ID
func (s *Service) Get(ctx context.Context, id string) (*memory.Memory, error) {
	return s.storage.Get(ctx, id)
}

// List lists memories by level
func (s *Service) List(ctx context.Context, level memory.MemoryLevel) ([]*memory.Memory, error) {
	opts := memory.ListOptions{FilterLevels: []memory.MemoryLevel{level}}
	return s.storage.List(ctx, opts)
}

// Delete deletes a memory
func (s *Service) Delete(ctx context.Context, id string) error {
	return s.storage.Delete(ctx, id)
}

// Search searches for similar memories
func (s *Service) Search(ctx context.Context, query string, level memory.MemoryLevel, topK int, minScore float64) ([]*memory.Memory, []float64, error) {
	embedding, err := s.embedder.Embed(ctx, query)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to embed query: %w", err)
	}

	results, err := s.storage.SearchAllLayers(ctx, embedding, memory.SearchOptions{
		TopK:         topK,
		MinScore:     minScore,
		FilterLevels: []memory.MemoryLevel{level},
	})
	if err != nil {
		return nil, nil, fmt.Errorf("failed to search: %w", err)
	}

	memories := make([]*memory.Memory, 0, len(results))
	scores := make([]float64, 0, len(results))
	for _, result := range results {
		memories = append(memories, result.Memory)
		scores = append(scores, result.Score)
	}

	return memories, scores, nil
}

func deriveTitle(content string) string {
	trimmed := strings.TrimSpace(content)
	if trimmed == "" {
		return "Memory"
	}
	line := strings.SplitN(trimmed, "\n", 2)[0]
	line = strings.TrimSpace(line)
	if len(line) < 3 {
		return "Memory"
	}
	const max = 60
	if len(line) > max {
		return strings.TrimSpace(line[:max]) + "..."
	}
	return line
}

// Helper functions

func mergeTags(existing, new []string) []string {
	tagSet := make(map[string]bool)
	for _, t := range existing {
		tagSet[t] = true
	}
	for _, t := range new {
		tagSet[t] = true
	}

	result := make([]string, 0, len(tagSet))
	for t := range tagSet {
		result = append(result, t)
	}
	return result
}

File: src/cortex/consolidation/service.py
"""Consolidation service — create or merge memories with deduplication."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime
from uuid import uuid4

import numpy as np

from cortex.embeddings.base import Embedder
from cortex.memory.service import MemoryService
from cortex.models.memory import Memory, MemoryContext, MemoryLevel, MemorySource
from cortex.models.results import ConsolidateResult
from cortex.search.cosine import normalize
from cortex.session import get_git_branch
from cortex.storage.base import SearchOptions, Storage
from cortex.utils import dedupe_list

_MERGE_SEPARATOR = "\n\n---\n\n"


@dataclass
class ConsolidateInput:
    """Input parameters for a consolidation operation."""

    synthesis: str  # Content to consolidate (becomes new/merged content)
    level: MemoryLevel
    title: str = ""
    tags: list[str] = field(default_factory=list)
    session_id: str = ""
    source: MemorySource = MemorySource.manual
    task_id: str = ""
    author: str = ""
    force: bool = False  # Skip duplicate detection when True
    # Save-context fields (PR #38)
    git_branch: str = ""
    agent_name: str = ""
    agent_session_id: str = ""
    user_prompt: str = ""


class ConsolidationService:
    """Creates or merges memories based on semantic similarity.

    When *force* is False (default):
    - Embeds the incoming synthesis
    - Searches for similar memories at the target level (threshold configurable, default 0.85)
    - If a match is found: merges content, updates embedding, preserves lineage
    - If no match: creates a new memory

    When *force* is True: always creates a new memory.
    """

    def __init__(
        self,
        storage: Storage,
        embedder: Embedder,
        similarity_threshold: float = 0.85,
    ) -> None:
        self._storage = storage
        self._embedder = embedder
        self._threshold = similarity_threshold

    def consolidate(self, inp: ConsolidateInput) -> ConsolidateResult:
        """Run the consolidation logic and return a result."""
        embed_text = MemoryService.EMBED_TEXT_TEMPLATE.format(
            title=inp.title or inp.synthesis[:60],
            content=inp.synthesis,
            tags=", ".join(inp.tags),
        )
        vector = self._embedder.embed(embed_text)

        if not inp.force:
            similar = self._find_similar(vector, inp.level, inp.session_id)
            if similar:
                existing, score = similar
                merged = self._merge(existing, inp, vector)
                return ConsolidateResult(
                    action="merged",
                    memory_id=merged.id,
                    level=str(merged.level),
                    similarity=score,
                    message=f"Merged with existing memory (score={score:.3f})",
                )

        # Create new memory
        # Auto-detect git branch when not explicitly provided (PR #38/#39)
        git_branch = inp.git_branch or get_git_branch()

        context = MemoryContext(
            session_id=inp.session_id,
            task_id=inp.task_id,
            author=inp.author,
            source=inp.source,
            timestamp=datetime.now(UTC),
            git_branch=git_branch,
            agent_name=inp.agent_name,
            agent_session_id=inp.agent_session_id,
            user_prompt=inp.user_prompt,
        )
        title = inp.title or Memory.derive_title(inp.synthesis)
        memory = Memory(
            id=str(uuid4()),
            level=inp.level,
            title=title,
            content=inp.synthesis,
            tags=inp.tags,
            embedding=vector,
            context=context,
        )
        self._storage.save(memory)
        return ConsolidateResult(
            action="created",
            memory_id=memory.id,
            level=str(memory.level),
            message="Created new memory",
        )

    def promote_to_semantic(self, memory_id: str) -> Memory:
        """Move an episodic memory to semantic, assigning a new ID and tracking origin."""
        source = self._storage.get(memory_id)
        # get() does not return the stored embedding; fetch it separately so the
        # promoted memory carries over the correct vector.
        embedding = self._storage.get_embedding(memory_id)
        new_memory = Memory(
            id=str(uuid4()),
            level=MemoryLevel.semantic,
            title=source.title,
            content=source.content,
            tags=source.tags,
            embedding=embedding,
            context=source.context,
            merged_from=[source.id],
        )
        self._storage.save(new_memory)
        # Mark the original as obsolete
        source.obsolete = True
        source.touch()
        self._storage.update(source)
        return new_memory

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _find_similar(
        self,
        vector: list[float],
        level: MemoryLevel,
        session_id: str,
    ) -> tuple[Memory, float] | None:
        """Return the most similar non-obsolete memory above threshold, or None."""
        opts = SearchOptions(
            top_k=1,
            min_score=self._threshold,
            filter_levels=[level],
            include_obsolete=False,
            session_id=session_id,
        )
        results = self._storage.search_all_layers(vector, opts)
        if results:
            return results[0].memory, results[0].score
        return None

    def _merge(self, existing: Memory, inp: ConsolidateInput, new_vector: list[float]) -> Memory:
        """Merge new content into an existing memory."""
        existing.content = existing.content + _MERGE_SEPARATOR + inp.synthesis
        existing.tags = dedupe_list(existing.tags + inp.tags)
        existing.merged_from = dedupe_list(existing.merged_from)
        existing.touch()

        # Average embeddings (fetch stored embedding since deserialized embedding is always [])
        stored_embedding = self._storage.get_embedding(existing.id)
        if stored_embedding and new_vector:
            averaged = np.mean([stored_embedding, new_vector], axis=0).tolist()
            existing.embedding = normalize(averaged)

        self._storage.update(existing)
        return existing
